src/wallet/wallet.cpp

```diff
@@ -3589,9 +3589,15 @@ void CWallet::CheckKeypoolMinSize() {
     unsigned int keypool_min = GetArg("-keypoolmin", DEFAULT_KEYPOOL_MIN);
     if (IsHDEnabled() && ((CanSupportFeature(FEATURE_HD_SPLIT) && setInternalKeyPool.size() < keypool_min) || (setExternalKeyPool.size() < keypool_min))) {
         // if the remaining keypool size is below the gap limit, shutdown
-        LogPrintf("%s: Keypool is too small. Shutting down. internal keypool: %d, external keypool: %d, keypool minimum: %d\n",
+        LogPrintf("%s: Number of keys in keypool is below minimum. Shutting down. internal keypool: %d, external keypool: %d, keypool minimum: %d\n",
                   __func__, setInternalKeyPool.size(), setExternalKeyPool.size(), keypool_min);
-        const std::string error_msg = "Keypool is too small. Shutting down";
+        const std::string error_msg = "Number of keys in keypool is below minimum and the wallet is encrytped. Bitcoin will now shutdown to avoid loss of funds.\n"
+                                      "This is probably because you are restoring an old backup wallet file which has not been topped up with the most recently "
+                                      "derived keys, and so you would not detect transactions involving those keys.\n"
+                                      "You can manually top-up your wallet keypool as follows:\n"
+                                      " - restart bitcoin with -keypoolmin set to 0\n"
+                                      " - unlock the wallet using walletpassphrase\n"
+                                      " - restart with -rescan. This will redownload the blockchain if you are running a pruned node.";
```

Comment:
This seems safe, but practically speaking shouldn't a manual rescan be unnecessary in most cases, because if the node shuts down when the keypool size is low, restarting should automatically rescan from `ReadBestBlock`.

Relatedly, maybe this PR should explicitly avoid calling `WriteBestBlock` when the keypool size is too low, in case shutdown takes a long time or -keypoolmin is set to 0 to recover.